nomad/structs/structs.go

```diff
@@ -2318,6 +2318,9 @@ type TaskState struct {
 	// The current state of the task.
 	State string
 
+	// Failed marks a task as having failed
+	Failed bool
```

Comment:
Alternatively would it offer any benefits to promote `State string` to an interface like:

```go
type State interface {
  Failed() bool
  Terminal() bool
  String() string
}

type runningState string

func (runningState) Terminal() bool { return false }
func (runningState) Failed() bool { return false }
func (r runningState) String() string { return string(r) }

type terminalState struct {
  reason string
  failed bool
}

func (terminalState) Terminal() bool { return false }
func (t terminalState) Failed() bool { return t.failed }
func (t terminalState) String() string { return t.reason }
```

Ok, after typing all that out maybe it is overkill. :) The only benefit I was aiming for what to prevent bugs where State+Failed get out of sync because they're managed independently. I'm not sure tons of boilerplate is worth protecting against that.